sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -858,6 +859,30 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+  * Push [[Limit]] operators through [[Join]] operators, iff the join type is outer joins.
+  * Adding extra [[Limit]] operators on top of the outer-side child/children.
+  */
+object PushLimitThroughOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
```

Comment:
If there is a sort, the plan is different. We will have a `Sort` below `Limit`. Thus, it is not applicable to this rule. 

For the full-outer join, I think the result is still correct. 

If the nodes below `limit` is deterministic, the result should be deterministic. Right? Sorry, this part is unclear to me. 

Based on my understanding, when we use `limit`, the users will not expect a deterministic results, unless they use `order by`.